lib/ABI/FunctionType/Layout.cpp

```diff
@@ -0,0 +1,963 @@
+  return ToRaw.convert(Function, Binary);
+}
+
+Layout::Layout(const model::CABIFunctionType &Function) {
+  const abi::Definition &ABI = abi::Definition::get(Function.ABI());
+  ToRawConverter Converter(ABI);
+
+  bool HasShadowPointerToAggregateReturnValue = false;
+  std::size_t CurrentStackOffset = 0;
+  const auto Architecture = model::ABI::getArchitecture(Function.ABI());
+  auto RV = Converter.distributeReturnValue(Function.ReturnType());
+  if (RV.SizeOnStack == 0) {
+    // Nothing on the stack, the return value fits into the registers.
+    auto &ReturnValue = ReturnValues.emplace_back();
+    ReturnValue.Type = Function.ReturnType();
+    ReturnValue.Registers = std::move(RV.Registers);
+  } else {
+    revng_assert(RV.Registers.empty(),
+                 "Register and stack return values should never be present "
+                 "at the same time.");
+
+    // Add an argument to represent the pointer to the return value location.
+    auto &RVLocationIn = Arguments.emplace_back();
+    RVLocationIn.Type = Function.ReturnType().getPointerTo(Architecture);
+    RVLocationIn.Kind = ArgumentKind::ShadowPointerToAggregateReturnValue;
+    HasShadowPointerToAggregateReturnValue = true;
+
+    if (ABI.ReturnValueLocationRegister() != model::Register::Invalid) {
+      // Return value is passed using the stack (with a pointer to the location
+      // in the dedicated register).
+      RVLocationIn.Registers.emplace_back(ABI.ReturnValueLocationRegister());
+    } else if (ABI.ReturnValueLocationOnStack()) {
+      // The location, where return value should be put in, is also communicated
+      // using the stack.
+      CurrentStackOffset += model::ABI::getPointerSize(ABI.ABI());
+      RVLocationIn.Stack = { 0, CurrentStackOffset };
+    } else {
+      revng_abort("Big return values are not supported by the current ABI");
+    }
+
+    // Also return the same pointer using normal means.
+    //
+    // NOTE: maybe some architectures do not require this.
+    // TODO: investigate.
+    auto &RVLocationOut = ReturnValues.emplace_back();
+    RVLocationOut.Type = RVLocationIn.Type;
+    revng_assert(RVLocationOut.Type.UnqualifiedType().isValid());
+
+    // To simplify selecting the register for it, use the full distribution
+    // routine again, but with the pointer instead of the original type.
+    auto RVOut = Converter.distributeReturnValue(RVLocationOut.Type);
+    revng_assert(RVOut.Size == model::ABI::getPointerSize(ABI.ABI()));
+    revng_assert(RVOut.Registers.size() == 1);
+    revng_assert(RVOut.SizeOnStack == 0);
+    RVLocationOut.Registers = std::move(RVOut.Registers);
+  }
+
+  auto Distributed = Converter.distributeArguments(Function.Arguments(),
+                                                   RV.SizeOnStack != 0);
+  revng_assert(Distributed.size() >= Function.Arguments().size());
+  std::size_t Index = 0;
+  for (const DistributedArgument &Argument : Distributed) {
+    if (!Argument.RepresentsPadding) {
```

Comment:
This is an internal representation, shared between `RFT` and `Layout`.

This flag does not affect `RFT`s but causes `Layout` to skip the argument entirely.

Reply:
> causes Layout to skip the argument entirely

My question was: why? Docs

Reply:
Because `Layout` doesn't need to be reversible. Consider the example
```cpp
void my_arm_function(int r0, __uint128_t r2_and_3);
```

In `RFT` we are forced to mark a "padding" register `r1` as argument - because we have no way of preserving the information that it has to stay untouched, in `Layout`, we can just say the truth and omit that information - hence we don't need the "padding" argument.

Reply:
OK, docs.

Reply:
How about something like this?
```cpp
/// The internal representation of the argument shared between both
/// the to-raw conversion and the layout.
struct DistributedArgument {
  /// A list of registers the argument uses.
  RegisterVector Registers = {};

  /// The total size of the argument (including padding if necessary) in bytes.
  std::uint64_t Size = 0;

  /// The size of the piece of the argument placed on the stack.
  /// \note: has to be equal to `0` or `Size` for any ABI for which
  ///        `abi::Definition::ArgumentsCanBeSplitBetweenRegistersAndStack()`
  ///        returns `false`.
  ///        Otherwise, it has to be an integer value `0 <= SizeOnStack <= Size`
  std::uint64_t SizeOnStack = 0;

  /// Mark this argument as a padding argument, which means an unused location
  /// (either a register or a piece of the stack) which needs to be seen as
  /// a separate argument to be able to place all the following arguments
  /// in the correct positions.
  ///
  /// The "padding" arguments are emitted as normal arguments in RawFunctionType
  /// but are omitted in `Layout`.
  bool RepresentsPadding = false;
};
```